src/NServiceBus.Core/Receiving/ReceiveComponent.cs

```diff
@@ -211,7 +221,67 @@ IPushMessages BuildMessagePump()
             return messagePumpFactory();
         }
 
-        ReceiveConfiguration configuration;
+        static void LoadMessageHandlers(Configuration configuration, List<Type> orderedTypes, IConfigureComponents container, IList<Type> availableTypes)
+        {
+            var types = new List<Type>(availableTypes);
+
+            foreach (var t in orderedTypes)
+            {
+                types.Remove(t);
+            }
+
+            types.InsertRange(0, orderedTypes);
+
+            ConfigureMessageHandlersIn(configuration, types, container);
+        }
+
+        static void ConfigureMessageHandlersIn(Configuration configuration, IEnumerable<Type> types, IConfigureComponents container)
+        {
+            var handlerRegistry = configuration.MessageHandlerRegistry ?? new MessageHandlerRegistry();
+
+            foreach (var t in types.Where(IsMessageHandler))
+            {
+                container.ConfigureComponent(t, DependencyLifecycle.InstancePerUnitOfWork);
+                handlerRegistry.RegisterHandler(t);
+            }
+
+            container.RegisterSingleton(handlerRegistry);
+        }
+
+        public static bool IsMessageHandler(Type type) // TODO: used in LoadMessageHandlersExtensions.ExecuteTheseHandlersFirst() extension method. Should it be moved elsewhere?
+        {
+            if (type.IsAbstract || type.IsGenericTypeDefinition)
+            {
+                return false;
+            }
+
+            return type.GetInterfaces()
+                .Where(@interface => @interface.IsGenericType)
+                .Select(@interface => @interface.GetGenericTypeDefinition())
+                .Any(genericTypeDef => genericTypeDef == IHandleMessagesType);
+        }
+
+        public class Configuration
+        {
+            public Configuration(SettingsHolder settings)
+            {
+                this.settings = settings;
+            }
+
+            public List<Type> ExecuteTheseHandlersFirst
+            {
+                get => settings.GetOrDefault<List<Type>>(ExecuteTheseHandlersFirstSettingKey);
+                set => settings.Set(ExecuteTheseHandlersFirstSettingKey, value);
+            }
+
+            public MessageHandlerRegistry MessageHandlerRegistry => settings.GetOrDefault<MessageHandlerRegistry>();
+
+            const string ExecuteTheseHandlersFirstSettingKey = "NServiceBus.ExecuteTheseHandlersFirst";
```

Comment:
Nitpick: private consts should go last in the class (below the settingsholder)

Reply:
Will change.
But can we automate this? editor.config? R#? It's really silly that we need to remember these. Especially when they are a matter of style and taste and not a compilation.